### ISSUE TITLE
some urls return descriptions that are way too long

### ISSUE THREAD

**Issue by maia** [gottfrois/link_thumbnailer#63]
I've encountered some urls that LinkThumbnailer can not correctly parse, and returns descriptions that either are just too long or also contain javascript and css. Here are some examples:

* http://www.pesterlloyd.net/html/1525homan.html
* http://www.oegfe.at/cms/index.php?id=143&tx_ttnews%5BbackPid%5D=141&tx_ttnews%5Btt_news%5D=897
* http://www.gemeinderecht.wien.at/vergabeportal/detailansicht.asp?atxtnr=37755#Ausschreibungsunterlagen_herunterladen
* https://www.detective.io/detective/the-migrants-files

In case there's no bug in LinkThumbnailer causing this behavior (I assume it's rather related to malformed html code), I'd like to suggest the enhancement of a config option that limits the length of title and description.

**Comment by gottfrois** [comment]
Hey, thanks for the links. I've been looking at this and I think there is room for improving the built-in algorithm that score the descriptions. In the meantime, you can easily add your own "graders" to handle such descriptions.

Here is a wiki page to explain how graders works https://github.com/gottfrois/link_thumbnailer/wiki/How-to-build-your-own-Grader%3F

Simply add your new grader in the configuration. For example in your case you might want something like this

```ruby
module LinkThumbnailer
  module Graders
    class Foo < ::LinkThumbnailer::Graders::Base
     
      # will substract 10 points for long description
      def call(current_score)
        text.length > 300 ? -10 : 0
      end
    end
  end
end

LinkThumbnailer.configure do |config|
 config.graders = [
    ->(description) { ::LinkThumbnailer::Graders::Length.new(description) },
    ->(description) { ::LinkThumbnailer::Graders::HtmlAttribute.new(description, :class) },
    ->(description) { ::LinkThumbnailer::Graders::HtmlAttribute.new(description, :id) },
    ->(description) { ::LinkThumbnailer::Graders::Position.new(description) },
    ->(description) { ::LinkThumbnailer::Graders::LinkDensity.new(description) },
    ->(description) { ::LinkThumbnailer::Graders::Foo.new(description) }
  ]
end
```

**Comment by gottfrois** [comment]
I greatly improved the description scoring strategy on the following branch https://github.com/gottfrois/link_thumbnailer/tree/improved_description_score

Can you give it a try and tell me how it behave for you please?

BTW nothing I can do for this website https://www.detective.io/detective/the-migrants-files/ since its their description `meta` attributes that define a huge description...

**Comment by maia** [comment]
I've tested it for ~200 urls and I've not encountered any "bad" descriptions. But I couldn't tell if the master branch would have done so.

The one thing I had to change is to remove the custom grader you suggested, if I remember correctly the method 'call' does not accept the current_score as argument any more. You might want to change the description accordingly.

As for the specific urls mentioned in the first post, I'm not worried about these. My app extracts parses tweets for urls and extracts information for these urls, so I'm dealing with a couple hundred new urls each day. These were simply the five urls with the longest descriptions in my db on that day.

Thanks!

**Comment by gottfrois** [comment]
Yes sorry about that, I should have told you. I basically refactored how graders works. Before there were simply doing a sum of positive or negative values but without any conventions regarding the "unit" used. Meaning that some graders could return `-1` or `1` value whereas other graders could return values order or magnitude different...

Anyway, now graders returns a probability (floats between 0 and 1) which is "the likely hood of a description to be considered good". Then probabilities are multiplied together to end up with a final computed probability for each descriptions.

When merging this into master, I will update the documentation as well.

I would like to thank you for testing this. May I ask you how you are using LinkThumbnailer in your business and what are your attempts regarding this gem? It may help me to better understand how people use it and focus on important features.

Also if you have time, I would be more than happy if you could write a small referral (couple lines) explaining how it helped you to solve what kind of issue. I would like to promote this tool a little more and would share those referrals on the readme page.

Thanks

**Comment by gottfrois** [comment]
released in `v3.0.0`
